src/packages/helpers/src/provider.ts

```diff
@@ -0,0 +1,126 @@
+import { Filter, PaginationOptions, BackendProvider } from '@exogee/graphweaver';
+
+export interface ProviderOptions<Entity, Context> {
+	init?(): Promise<Context>;
+	create?(context: Context, entity: Partial<Entity>): Promise<Entity>;
+	read(
+		context: Context,
+		filter: Filter<Entity>,
+		pagination?: PaginationOptions
+	): Promise<Entity | Array<Entity> | null>;
+	update?(context: Context, id: string, entity: Partial<Entity>): Promise<Entity>;
+	remove?(context: Context, filter: Filter<Entity>): Promise<boolean>;
+	backendId: string;
+}
+
+export const createProvider = <Entity, Context>(options: ProviderOptions<Entity, Context>) => {
+	class Provider<Entity, Context> implements BackendProvider<Entity, Entity> {
+		readonly backendId: string;
+
+		create: ProviderOptions<Entity, Context>['create'];
+		read: ProviderOptions<Entity, Context>['read'];
+		update: ProviderOptions<Entity, Context>['update'];
+		remove: ProviderOptions<Entity, Context>['remove'];
+		initFn: Promise<void>;
+
+		context: Context | undefined;
+
+		constructor({
+			create,
+			read,
+			update,
+			remove,
+			init,
+			backendId,
+		}: ProviderOptions<Entity, Context>) {
+			this.backendId = backendId;
+			this.create = create;
+			this.read = read;
+			this.update = update;
+			this.remove = remove;
+			this.initFn = new Promise<void>((resolve, reject) => {
+				if (!init) {
+					resolve();
+					return;
+				}
+				init().then((context) => {
+					this.context = context;
+					resolve();
+				});
+			});
+		}
+
+		init() {
+			return this.initFn;
+		}
+
+		async find(filter: Filter<Entity>, pagination?: PaginationOptions): Promise<Array<Entity>> {
+			await this.init;
+			const result = await this.read(this.context as Context, filter, pagination);
+			return result === null ? [] : Array.isArray(result) ? result : [result];
+		}
+
+		async findOne(filter: Filter<Entity>): Promise<Entity | null> {
```

Comment:
should findOne just call find with a limit of 1?

Reply:
seems reasonable